fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -126,26 +126,37 @@ class FullyShardedDataParallel(nn.Module):
         mixed_precision (bool, Optional):
             if ``True``, inputs, activations and gradients will be kept in FP16;
             computation and communication will occur in FP16; and a (sharded)
-            master copy of the model weights will be maintained in FP32.
-        fp32_reduce_scatter (bool, Optional):
-            if ``True``, then reduce-scatter gradients in FP32. This is only
-            relevant when *``mixed_precision``* is ``True``.
+            master copy of the model weights will be maintained in FP16.
+            Weight, buffer and communication data types can be overriden using
+            the 3 options below.
+            Note, this is expected to work with autocast context from pytorch.
+            Note, some layers, e.g. convolutions, may only work with
+            ``fp32_compute_dtype==True`` below.
+            Default: False
+        fp32_compute_dtype (bool, Optional):
+            if ``True``, use FP32 for full parameters for computation.
+            Default: None, which means default to FP16 or FP32 based on the
+            *``mixed_precision``* argument.
+        fp32_buffer_dtype (bool, Optional):
+            if ``True``, use FP32 for buffers for computation.
+            Default: None, which means using the value of the
+            *``fp32_compute_dtype``* argument.
+        fp32_reduce_scatter (bool):
+            if ``True``, then reduce-scatter gradients in FP32.
+            This is only relevant when *``mixed_precision``* is ``True``.
+            Default: False
         flatten_parameters (bool, Optional):
             if ``True``, flatten parameters into a single contiguous tensor,
             which improves training speed.
+            Default: True
         cpu_offload (bool, Optional):
-            if ``True``, offload FP32 params to CPU. This is only relevant when
-            *``mixed_precision``* is ``True``.
-        compute_dtype (torch.dtype, Optional):
```

Comment:
reflecting on this, I think we should still support compute_dtype, and potentially have fp32_compute_dtype/buffer_dtype as convenience flags instead. The reason is that the latest A100 GPUs support torch.bfloat16 and torch.tfloat32, which are actually compatible with FSDP if you do `FSDP(mixed_precision=True, compute_dtype=torch.bfloat16)`

Reply:
Good point. Let me add them back.

Reply:
I thought about this more, should we do something like the following?

compute_dtype, buffer_dtype, reduce_scatter_dtype all take the same data type input, i.e. Union[torch.dtype, str]. The reason for str is that for easier config where users can supply string to be converted to the right dtype (like "fp16" -> torch.float16, etc.).

What do you think?

Reply:
Yeah, seems cleaner too, that way we don't have redundant kwargs.

There seems to have been some discussion about this in a PyTorch issue: https://github.com/pytorch/pytorch/issues/40471, but sadly no public interface. We can easy enumerate the dtypes we support: float16, bfloat16, float32. I was wrong about tensorfloat32, it's not actually a separate dtype, but we should support at least the other three.